### ISSUE TITLE
 Unhandled exception from module 'dashboard' while running on mgr.a: [('x509 certificate routines', 'X509_check_private_key', 'key values mismatch')]

### ISSUE THREAD

**Issue by melnik13** [rook/rook#4207]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
The dashboard module stopped working after a restart.

**Expected behavior:**
The dashboard module should keep working.

**How to reproduce it (minimal and precise):**
It's hard to say how to reproduce it. I restarted the pods of rook-ceph multiple times, but it didn't come to that result before.

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
```
apiVersion: v1
items:
- apiVersion: ceph.rook.io/v1
  kind: CephCluster
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"ceph.rook.io/v1","kind":"CephCluster","metadata":{"annotations":{},"name":"rook-ceph","namespace":"rook-ceph"},"spec":{"annotations":null,"cephVersion":{"allowUnsupported":false,"image":"ceph/ceph:v14.2.2-20190826"},"configOverrides":[{"option":"debug ms","value":"1/5","who":"global"},{"option":"debug_osd","value":"10","who":"osd.0"},{"option":"debug-mon","value":"10","who":"mon.a"}],"dashboard":{"enabled":true},"dataDirHostPath":"/var/lib/rook","disruptionManagement":{"managePodBudgets":false,"osdMaintenanceTimeout":30},"mon":{"allowMultiplePerNode":false,"count":3},"monitoring":{"enabled":false,"rulesNamespace":"rook-ceph"},"network":{"hostNetwork":false},"rbdMirroring":{"workers":0},"resources":null,"storage":{"config":null,"deviceFilter":"^vdc","location":null,"useAllDevices":true,"useAllNodes":true}}}
    creationTimestamp: "2019-09-01T07:41:41Z"
    finalizers:
    - cephcluster.ceph.rook.io
    generation: 48589
    name: rook-ceph
    namespace: rook-ceph
    resourceVersion: "37795851"
    selfLink: /apis/ceph.rook.io/v1/namespaces/rook-ceph/cephclusters/rook-ceph
    uid: 5ed92ec1-ed0e-4963-89e9-a3b3e795f7e1
  spec:
    cephVersion:
      image: ceph/ceph:v14.2.2-20190826
    dashboard:
      enabled: true
    dataDirHostPath: /var/lib/rook
    disruptionManagement: {}
    external:
      enable: false
    mon:
      count: 3
    monitoring:
      rulesNamespace: rook-ceph
    network:
      hostNetwork: false
      provider: ""
      selectors: null
    rbdMirroring:
      workers: 0
    storage:
      config: null
      deviceFilter: ^vdc
      storageClassDeviceSets: null
      useAllDevices: true
      useAllNodes: true
  status:
    ceph:
      details:
        MGR_MODULE_ERROR:
          message: 'Module ''dashboard'' has failed: [(''x509 certificate routines'',
            ''X509_check_private_key'', ''key values mismatch'')]'
          severity: HEALTH_ERR
      health: HEALTH_ERR
      lastChanged: "2019-10-27T20:20:02Z"
      lastChecked: "2019-10-29T08:02:59Z"
      previousHealth: HEALTH_WARN
    state: Created
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

* Operator's logs, if necessary
The operator's log file contains the following lines:
```
2019-10-29 07:48:34.204691 I | exec: Running command: ceph mgr module enable dashboard --force --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.
config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json --out-file /tmp/097468789
2019-10-29 07:48:35.256499 I | exec: module 'dashboard' is already enabled
2019-10-29 07:48:40.264490 I | op-mgr: the dashboard secret was already generated
2019-10-29 07:48:40.264623 I | exec: Running command: ceph dashboard create-self-signed-cert --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.co
nfig --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json --out-file /tmp/529309264
2019-10-29 07:48:40.914502 I | exec: Error EIO: Module 'dashboard' has experienced an error and cannot handle commands: [('x509 certificate routines', 'X509_check_private_key', 'key values mismatch')]
2019-10-29 07:48:40.914612 I | op-mgr: dashboard is already initialized with a cert
2019-10-29 07:48:40.914688 I | exec: Running command: ceph config get mgr.a mgr/dashboard/url_prefix --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json --out-file /tmp/234799471
2019-10-29 07:48:41.674014 I | exec: Running command: ceph config rm mgr.a mgr/dashboard/url_prefix --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json --out-file /tmp/672870914
2019-10-29 07:48:42.490684 I | exec: Running command: ceph config get mgr.a mgr/dashboard/server_port --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json --out-file /tmp/634351481
2019-10-29 07:48:43.216459 I | exec: Running command: ceph config set mgr.a mgr/dashboard/server_port 8443 --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json --out-file /tmp/181432452
2019-10-29 07:48:44.006458 I | exec: Running command: ceph config get mgr.a mgr/dashboard/ssl --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json --out-file /tmp/283686419
2019-10-29 07:48:44.955481 I | exec: Running command: ceph config rm mgr.a mgr/dashboard/ssl --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json --out-file /tmp/935099990
2019-10-29 07:48:45.833807 I | op-mgr: dashboard config has changed
2019-10-29 07:48:45.833833 I | op-mgr: restarting the mgr module
2019-10-29 07:48:45.833898 I | exec: Running command: ceph mgr module disable dashboard --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json --out-file /tmp/011189181
2019-10-29 07:48:46.940875 I | exec: Running command: ceph mgr module enable dashboard --force --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json --out-file /tmp/455280632
2019-10-29 07:48:48.112332 I | op-mgr: dashboard service already exists
2019-10-29 07:48:48.162367 I | op-mgr: mgr metrics service already exists
```

The mgr's log-file contains that:
```
debug 2019-10-29 07:51:26.416 7fe56302a700 -1 log_channel(cluster) log [ERR] : Unhandled exception from module 'dashboard' while running on mgr.a: [('x509 certificate routines', 'X509_check_private_key', 'key values mismatch')]
debug 2019-10-29 07:51:26.416 7fe56302a700 -1 dashboard.serve:
debug 2019-10-29 07:51:26.418 7fe566831700  1 mgr load Constructed class from module: prometheus
debug 2019-10-29 07:51:26.420 7fe566831700  1 mgr load Constructed class from module: restful
debug 2019-10-29 07:51:26.420 7fe566831700  1 mgr load Constructed class from module: rook
debug 2019-10-29 07:51:26.422 7fe566831700  1 mgr load Constructed class from module: status
debug 2019-10-29 07:51:26.416 7fe56302a700 -1 Traceback (most recent call last):
  File "/usr/share/ceph/mgr/dashboard/module.py", line 353, in serve
    uri = self.await_configuration()
  File "/usr/share/ceph/mgr/dashboard/module.py", line 259, in await_configuration
    uri = self._configure()
  File "/usr/share/ceph/mgr/dashboard/module.py", line 225, in _configure
    context.use_privatekey_file(pkey_fname, crypto.FILETYPE_PEM)
Error: [('x509 certificate routines', 'X509_check_private_key', 'key values mismatch')]

debug 2019-10-29 07:51:26.423 7fe566831700  1 mgr load Constructed class from module: volumes
debug 2019-10-29 07:51:26.424 7fe55ffe4700  1 mgr[restful] server not running: no certificate configured
```

**Environment**:
* OS (e.g. from /etc/os-release):
`CentOS Linux release 7.6.1810 (Core)` on all the nodes
* Kernel (e.g. `uname -a`):
`Linux /*different*/.c13.tuchakube.tucha13.net 3.10.0-957.21.3.el7.x86_64 #1 SMP Tue Jun 18 16:35:19 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux` on all the nodes
* Cloud provider or hardware configuration:
There are 4 nodes and 2 masters sharing their own private network.
* Rook version (use `rook version` inside of a Rook Pod):
`rook: v1.0.0-597.g5c59a1e`
* Storage backend version (e.g. for ceph do `ceph -v`):
`ceph version 14.2.2 (4f8fa0a0024755aae7d95567c63f11d6862d55be) nautilus (stable)`
* Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.2", GitCommit:"f6278300bebbb750328ac16ee6dd3aa7d3549568", GitTreeState:"clean", BuildDate:"2019-08-05T09:23:26Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.1", GitCommit:"4485c6f18cee9a5d3c3b4e523bd27972b1b53892", GitTreeState:"clean", BuildDate:"2019-07-18T09:09:21Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
```
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
Bare metal
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):
`HEALTH_ERR Module 'dashboard' has failed: [('x509 certificate routines', 'X509_check_private_key', 'key values mismatch')]`

I also have the following group of secrets in the `rook-ceph` namespace:
```
NAME                                             TYPE                                  DATA   AGE
default-token-p74vf                              kubernetes.io/service-account-token   3      57d
rgw-tls                                          kubernetes.io/tls                     3      23d
rgw-tls-secret                                   Opaque                                1      56d
rook-ceph-admin-keyring                          kubernetes.io/rook                    1      57d
rook-ceph-cmd-reporter-token-kgsng               kubernetes.io/service-account-token   3      57d
rook-ceph-config                                 kubernetes.io/rook                    2      57d
rook-ceph-csi                                    kubernetes.io/rook                    4      57d
rook-ceph-dashboard-password                     kubernetes.io/rook                    1      57d
rook-ceph-mds-cephfs-a-keyring                   kubernetes.io/rook                    1      57d
rook-ceph-mds-cephfs-b-keyring                   kubernetes.io/rook                    1      57d
rook-ceph-mgr-a-keyring                          kubernetes.io/rook                    1      57d
rook-ceph-mgr-token-tk7k6                        kubernetes.io/service-account-token   3      57d
rook-ceph-mon                                    kubernetes.io/rook                    4      57d
rook-ceph-mons-keyring                           kubernetes.io/rook                    1      57d
rook-ceph-object-user-rgw-demo2019072213000477   kubernetes.io/rook                    2      56d
rook-ceph-osd-token-4dw67                        kubernetes.io/service-account-token   3      57d
rook-ceph-rgw-rgw-a-keyring                      kubernetes.io/rook                    1      56d
rook-ceph-system-token-87gkh                     kubernetes.io/service-account-token   3      57d
rook-csi-cephfs-plugin-sa-token-r6fsq            kubernetes.io/service-account-token   3      57d
rook-csi-cephfs-provisioner-sa-token-nrgf4       kubernetes.io/service-account-token   3      57d
rook-csi-rbd-plugin-sa-token-r9bpj               kubernetes.io/service-account-token   3      57d
rook-csi-rbd-provisioner-sa-token-545qs          kubernetes.io/service-account-token   3      57d
```

**Is it really a bug?**
Why I marked this issue as a bug? As I think, the operator should resolve such issue by itself, so maybe some additional checks and scenarios should be implemented.

**Any suggestions on how to resolve it?**
Should I try to remove some of the secrets and hope that the operator will recreate them? Thanks in advance for any ideas!

**Comment by danilo404** [comment]
Hi @melnik13. You've probably found the same work around by now but just in case and for future reference. I had the exact same problem after performing node restarts in the cluster and the solution that worked for me was deleting the key from the ceph configuration kv store and creating the certificates again. I did this from the ceph tools pod:
```
ceph config-key rm mgr/dashboard/key
ceph config-key rm mgr/dashboard/crt
ceph dashboard create-self-signed-cert
```

**Comment by melnik13** [comment]
> Hi @melnik13. You've probably found the same workaround by now but just in case and for future reference. I had the exact same problem after performing node restarts in the cluster and the solution that worked for me was deleting the key from the ceph configuration kv store and creating the certificates again.

Thank you, @danilo404, somewhy I didn't see your comment, that's why I hadn't sent you my thanks for your attention.

**Comment by FrelDX** [comment]
> Hi @melnik13. You've probably found the same workaround by now but just in case and for future reference. I had the exact same problem after performing node restarts in the cluster and the solution that worked for me was deleting the key from the ceph configuration kv store and creating the certificates again. I did this from the ceph tools pod:
> 
> ```
> ceph config-key rm mgr/dashboard/key
> ceph config-key rm mgr/dashboard/crt
> ceph dashboard create-self-signed-cert
> ```

[root@rook-ceph-tools-fff8ccb89-qpqbr /]# ceph config-key rm mgr/dashboard/key
key deleted
[root@rook-ceph-tools-fff8ccb89-qpqbr /]# ceph config-key rm mgr/dashboard/crt
key deleted
[root@rook-ceph-tools-fff8ccb89-qpqbr /]# ceph dashboard create-self-signed-cert
Error EIO: Module 'dashboard' has experienced an error and cannot handle commands: [('x509 certificate routines', 'X509_check_private_key', 'key values mismatch')]

**Comment by dominic-p** [comment]
I'm running into this same thing. All of a sudden I can't schedule pods that need persistent volumes:

> Warning  FailedScheduling  31m   default-scheduler  0/3 nodes are available: 3 pvc(s) bound to non-existent pv(s).

When I run `ceph status` I get:

```
    health: HEALTH_ERR
            Module 'dashboard' has failed: [('x509 certificate routines', 'X509_check_private_key', 'key values mismatch')]
```

I'm not sure if it's related, but I'm getting the same thing as @FrelDX

**Comment by huangming** [comment]
I'm getting the same thing. After some steps everything is ok:
1. change ssl: true to false on cluster.yaml and apply it.
```
2. 
ceph config-key rm mgr/dashboard/key
ceph config-key rm mgr/dashboard/crt
ceph dashboard create-self-signed-cert
```
3. change ssl: false to true on cluster.yaml and apply it.

**Comment by voarsh2** [comment]
I've recently ran into this on Rook v1.11.8. Will check this issue over and see if I can resolve it.

**Comment by kuroyi** [comment]
I recently had a node fail in my Rook cluster and replaced it with a new node. I went through the steps to remove the old OSD's, etc, but I was still dealing with seemingly random Rook issues causing it to essentially keep restarting some pods. One of the issues was this dashboard cert error. I don't know why it occurred but after using the procedure outlined by @huangming to recreate the cert it also resolved the other Rook issues I was having. Or maybe I fixed everything else is this was the last domino; I'm not sure.

    health: HEALTH_ERR
            Module 'dashboard' has failed: [('x509 certificate routines', 'X509_check_private_key', 'key values mismatch')]
            mons ag,am are using a lot of disk space
            1/3 mons down, quorum ag,am
            21 daemons have recently crashed
